pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -41,94 +41,147 @@ var client = http.Client{
-
 	switch fas.Spec.Policy.Type {
 	case autoscalingv1.BufferPolicyType:
 		return applyBufferPolicy(fas.Spec.Policy.Buffer, f)
 	case autoscalingv1.WebhookPolicyType:
 		return applyWebhookPolicy(fas.Spec.Policy.Webhook, f)
 	}
 
-	return f.Status.Replicas, false, errors.New("wrong policy type, should be one of: Buffer, Webhook")
+	return 0, false, errors.New("wrong policy type, should be one of: Buffer, Webhook")
 }
 
-func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (int32, bool, error) {
-	faReq := autoscalingv1.FleetAutoscaleReview{
-		Request: &autoscalingv1.FleetAutoscaleRequest{
-			UID:       uuid.NewUUID(),
-			Name:      f.Name,
-			Namespace: f.Namespace,
-			Status:    f.Status,
-		},
-		Response: nil,
-	}
-	b, err := json.Marshal(faReq)
-	urlStr := ""
+func buildURLFromWebhookPolicy(w *autoscalingv1.WebhookPolicy) (*url.URL, error) {
+	var u *url.URL
+	var err error
+
 	if w.URL != nil {
-		urlStr = *w.URL
-	}
-	var faResp autoscalingv1.FleetAutoscaleReview
-	servicePath := ""
-	if w.Service != nil {
-		if w.Service.Path != nil {
-			servicePath = *w.Service.Path
+		if *w.URL == "" {
+			return nil, errors.New("URL was not provided")
 		}
+
+		u, err = url.ParseRequestURI(*w.URL)
 		if err != nil {
-			return f.Status.Replicas, false, err
+			return nil, err
+		}
+	} else {
+		if w.Service.Name == "" {
+			return nil, errors.New("service name was not provided")
+		}
+
+		var servicePath string
+		if w.Service.Path != nil {
+			servicePath = *w.Service.Path
 		}
 
 		if w.Service.Namespace == "" {
 			w.Service.Namespace = "default"
 		}
-		scheme := "http://"
+
+		scheme := "http"
 		if w.CABundle != nil {
-			scheme = "https://"
+			scheme = "https"
+
+			err = setCABundle(w.CABundle)
+			if err != nil {
+				return nil, err
+			}
 		}
-		urlStr = fmt.Sprintf("%s%s.%s.svc:8000/%s", scheme, w.Service.Name, w.Service.Namespace, servicePath)
+
+		u = createURL(scheme, w.Service.Name, w.Service.Namespace, servicePath)
 	}
-	if urlStr == "" {
-		return f.Status.Replicas, false, errors.New("URL was not provided")
+
+	return u, nil
+}
+
+func createURL(scheme, name, namespace, path string) *url.URL {
+	return &url.URL{
+		Scheme: scheme,
+		Host:   fmt.Sprintf("%s.%s.svc:8000", name, namespace),
+		Path:   path,
 	}
+}
 
-	u, err := url.Parse(urlStr)
+func setCABundle(CABundle []byte) error {
+	// We can have multiple fleetautoscalers with different CABundles defined,
+	// so we switch client.Transport before each POST request
+	rootCAs := x509.NewCertPool()
+	if ok := rootCAs.AppendCertsFromPEM(CABundle); !ok {
+		return errors.New("no certs were appended from caBundle")
+	}
+	client.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: rootCAs,
+		},
+	}
+	return nil
+}
+
+func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (replicas int32, limited bool, err error) {
```

Comment:
why does this function use named return parameters?

Reply:
I did that in order to catch and wrap an error in a defer() statement in case of an error after res.Body.Close()